Userland/Libraries/LibCore/EventLoopImplementationUnix.cpp

```diff
@@ -39,18 +40,222 @@ bool has_flag(int value, int flag)
+        }
+    }
+
+    void heapify_down(size_t i)
+    {
+        while (i * 2 + 1 < m_heap.size()) {
+            size_t min_child = i * 2 + 1;
+            size_t other_child = i * 2 + 2;
+            if (other_child < m_heap.size() && heap_compare(other_child, min_child))
+                min_child = other_child;
+            if (heap_compare(i, min_child))
+                break;
+            intrusive_swap(m_heap[i], m_heap[min_child]);
+            i = min_child;
+        }
+    }
+
+    void heap_add(EventLoopTimeout* timeout)
+    {
+        m_heap.append(timeout);
+        timeout->set_index({}, static_cast<ssize_t>(m_heap.size() - 1));
+        heapify_up(m_heap.size() - 1);
+    }
+
+    void heap_pop(size_t i)
+    {
+        while (i != 0) {
+            intrusive_swap(m_heap[i], m_heap[(i - 1) / 2]);
+            i = (i - 1) / 2;
+        }
+        intrusive_swap(m_heap[0], m_heap[m_heap.size() - 1]);
+        (void)m_heap.take_last();
+        heapify_down(0);
+    }
+
+    Vector<EventLoopTimeout*, 8> m_heap;
+    Vector<EventLoopTimeout*, 8> m_scheduled_timeouts;
+};
+
+class EventLoopTimer final : public EventLoopTimeout {
+public:
+    static constexpr auto delay_tolerance = Duration::from_milliseconds(5);
+
+    EventLoopTimer() = default;
+
+    void reload(MonotonicTime const& now) { m_fire_time = now + interval; }
+
+    virtual void fire(TimeoutSet& timeout_set, MonotonicTime current_time) override
+    {
+        auto strong_owner = owner.strong_ref();
+
+        if (!strong_owner)
+            return;
+
+        if (should_reload) {
+            MonotonicTime next_fire_time = m_fire_time + interval;
+            if (next_fire_time <= current_time) {
+                auto delay = current_time - next_fire_time;
+                if (delay >= delay_tolerance && !interval.is_zero()) {
+                    auto iterations = delay.to_milliseconds() / max<i64>(1, interval.to_milliseconds()) + 1;
+                    dbgln("Can't keep up! Skipping approximately {} iteration(s) of a reloading timer (delayed by {}ms).", iterations, delay.to_milliseconds());
+                }
+                next_fire_time = current_time + interval;
+            }
+            m_fire_time = next_fire_time;
+            if (next_fire_time != current_time) {
+                timeout_set.schedule_absolute(this);
+            } else {
+                // NOTE: Unfortunately we need to treat timeouts with the zero interval in a
+                //       special way. TimeoutSet::schedule_absolute for them will result in an
+                //       infinite loop. TimeoutSet::schedule_relative, on the other hand, will do a
+                //       correct thing of scheduling them for the next iteration of the loop.
+                m_duration = {};
+                timeout_set.schedule_relative(this);
+            }
+        }
+
+        // FIXME: This is just bad design. There aren't any notifications for when an EventReciever
```

Comment:
This was created to limit event loop churn when a window is not visible. So, if you have a window fully occluded behind another window, none of the timers in the process that have this flag would fire. I believe the canonical example of this was watching the CPU graph go to zero when the Fire Demo application was hidden behind another window. The intent is that repetitive timers would skip many many cycles during the period that the window is occluded, and then resume when the window/widget is visible again.

Reply:
My point is that while we skip doing most of the CPU-heavy work, we still wake every 16ms to run an if statement. IMO, a better way to handle this is just to unregister the timer while it is not needed. This way the event loop will properly sleep.

Reply:
gotcha. Perhaps that suggested solution should be part of the comment? I didn't get that implication reading it the first time.